doc/erofs.md

```diff
@@ -0,0 +1,431 @@
+    the kernel doesn't know about any features which might be present.  The
+    flags:
+    -   `SB_CHKSUM` (`0x0001`): set if the checksum field in the superblock is
+        populated.  Otherwise, the checksum is ignored.
+    -   `MTIME` (`0x0002`): at first, erofs named the timestamp fields `ctime`
+        instead of `mtime`.  That got changed a long time ago, and this flag
+        got added to indicate filesystems that were created with the new
+        semantics. This flag has absolutely zero impact at run time: the kernel
+        ignores it.
+    -   `XATTR_FILTER` (`0x0004`): set if the xattr bloom filter should be
+        used.  Read about this in the inode section.
+*   `blkszbits`: log2 of the block size.  Better set this to 12 (4096).
+*   `root_nid`: the reference to the root inode.  See the inodes section for
+    what that means.  Normally inodes are stored in u64, but this is somewhat
+    randomly a u16, which means that you're gonna need to put the root
+    directory near the start.
+*   `inos`: the total number of inodes defined.  This is only used for
+    `statfs()` purposes.
+*   `build_time`, `build_time_nsec`: this is something like a compression
+    feature if you want all (or many) files in your filesystem to have the same
+    mtime.  Then you can use the "compact" inode layout, which doesn't have its
+    own `mtime` field, and this one will be used instead.  If you don't have
+    compact inodes then this is meaningless.
+*   `blocks`: total filesystem block size.  This is only used for `statfs()`.
+*   `meta_blkaddr`: the start of the "metadata area".  This is where the inodes
+    are.  This is a block address, so it gets multiplied by the block size to
+    determine the actual offset.
+*   `xattr_blkaddr`: the start of the "shared xattr area".  See the "Shared
+    xattr" and "Inodes" sections for more info.
+
+## Extended attributes
+
+There are two options for storing xattr data in a erofs:
+*   inline with the inode itself
+*   in a "shared xattr" struct somewhere
+
+The format of both of these is the same.
+
+The inline thing is nice and simple, but it might be space-inefficient for
+cases where the same (key, value) pair appears over and over again (which might
+be the case for things like security labels and acls and the like).
+
+### Prefix indexes
+
+A rudimentary form of compression is supported on xattr names.  There are a
+number of hardcoded "common prefixes" defined with the `EROFS_XATTR_INDEX_`
+constants in `erofs_fs.h`.  Confusingly, although `LUSTRE` is present, it's not
+wired up in the kernel.  Don't use that one.
+
+The basic idea is that you find the prefix for your xattr from the list (like
+`user.` or `security.`) and then you store only the "suffix" part, along with
+the prefix index.  If you can't find a prefix, you use 0 (which is conceptually
+a prefix of "").  If the prefix matches the entire name then the suffix is `""`.
```

Comment:
We could define `0` as a valid prefix number, although 0 will be ignored and cannot be listed at all in kernel.

Reply:
That's good to know.  I think I didn't fully understand how the "prefix" part of xattrs is so special when I was writing this...